jcapi-systems.go

```diff
@@ -0,0 +1,334 @@
+	if err != nil {
+		return nil, fmt.Errorf("ERROR: Post to JumpCloud failed, err='%s'", err)
+	}
+
+	returnVal = getJCSystemsFromInterface(jcSystemRec)
+
+	if withTags {
+		tags, err := jc.GetAllTags()
+		if err != nil {
+			return nil, fmt.Errorf("ERROR: Could not get tags, err='%s'", err)
+		}
+
+		for idx, _ := range returnVal {
+			returnVal[idx].AddJCTagsToSystem(tags)
+		}
+	}
+
+	return returnVal, nil
+}
+
+func (jc JCAPI) GetSystemById(systemId string, withTags bool) (system JCSystem, err JCError) {
+	url := fmt.Sprintf("/systems/%s", systemId)
+
+	retVal, err := jc.Get(url)
+	if err != nil {
+		err = fmt.Errorf("ERROR: Could not get system by ID '%s', err='%s'", systemId, err)
+	}
+
+	if retVal != nil {
+		getJCSystemFieldsFromInterface(retVal.(map[string]interface{}), &system)
+
+		if withTags {
+			// I should be able to use err below as the err return value, but there's
+			// a compiler bug here in that it thinks a := of err is shadowed here,
+			// even though tags should be the only variable declared with the :=
+			tags, err2 := jc.GetAllTags()
+			if err != nil {
+				err = fmt.Errorf("ERROR: Could not get tags, err='%s'", err2)
+				return
+			}
+
+			system.AddJCTagsToSystem(tags)
+		}
+	}
+
+	return
+}
+
+func (jc JCAPI) GetSystems(withTags bool) ([]JCSystem, JCError) {
+	var returnVal []JCSystem
+
+	for skip := 0; skip == 0 || len(returnVal) == searchLimit; skip += searchSkipInterval {
+		url := fmt.Sprintf("/systems?sort=hostname&skip=%d&limit=%d", skip, searchLimit)
+
+		jcSysRec, err2 := jc.Get(url)
+		if err2 != nil {
+			return nil, fmt.Errorf("ERROR: Get to JumpCloud failed, err='%s'", err2)
+		}
+
+		returnVal = getJCSystemsFromInterface(jcSysRec)
+
+		if withTags {
+			tags, err := jc.GetAllTags()
+			if err != nil {
+				return nil, fmt.Errorf("ERROR: Could not get tags, err='%s'", err)
+			}
+
+			for idx, _ := range returnVal {
+				returnVal[idx].AddJCTagsToSystem(tags)
+			}
+		}
+
+	}
+	return returnVal, nil
+}
+
+//
+// Update a system
+//
+func (jc JCAPI) UpdateSystem(system JCSystem) (systemId string, err JCError) {
+	data, err := json.Marshal(system)
+	if err != nil {
+		return "", fmt.Errorf("ERROR: Could not marshal JCSystem object, err='%s'", err)
+	}
+	url := "/systems/" + system.Id
+
+	jcSysRec, err := jc.Do(MapJCOpToHTTP(Update), url, data)
+	if err != nil {
+		return "", fmt.Errorf("ERROR: Could not update JCSystem object, err='%s'", err)
+	}
+	//TODO it looks like this returns a partial system record?
+	//TODO i need to understand this better
+	var returnSystem JCSystem
+	getJCSystemFieldsFromInterface(jcSysRec.(map[string]interface{}), &returnSystem)
+	//parsedKeys := jcSysRec.(map[string]interface{})
+	//systemId = parsedKeys["_id"].(string)
+
+	if returnSystem.Id != system.Id {
+		return "", fmt.Errorf("ERROR: JumpCloud did not return the same ID - this should never happen!")
+	}
+	systemId = returnSystem.Id
+	return systemId, nil
+}
+
+func (jc JCAPI) DeleteSystem(system JCSystem) JCError {
```

Comment:
Could you add a comment here stating that this will cause JumpCloud to uninstall the agent on this system, and will lose control of the system after the call returns?